### ISSUE TITLE
Custom renderer for exceptions

### ISSUE THREAD

**Issue by Giorgi** [NLog/NLog#3519]
**NLog version**: 4.6.6

**Platform**: All

**Current NLog config** (xml or C#, if relevant)

```xml
<nlog>
  <targets>
    <target xsi:type="File" name="f" fileName="${basedir}/logs/${shortdate}.log"
            layout="${longdate} ${uppercase:${level}} ${message} ${exception:format=message}" />
  </targets>
  <rules>
    <logger name="*" minlevel="Debug" writeTo="f" />
  </rules>
</nlog>
```

Most of the exception subclasses have custom properties which are very useful when diagnosing the cause of the exception. For example `Win32Exception` has a `NativeErrorCode` property. If we want the property to get logged we need to add a catch block specifically for `Win32Exception` and log ` NativeErrorCode`. This can lead to useless and duplicated catch blocks.

I propose adding a way to create custom renderers for specific exception subclasses so the instances of that exception type will be formatted with that custom renderer. Here is an example for registering custom exception renderer for `Win32Exception` which will log `NativeErrorCode` in addition to default data logged for every exception:

```
ExceptionLayoutRenderer.RegisterExceptionRenderer<Win32Exception>(exception => $"NativeErrorCode: {exception.NativeErrorCode}");
```

With this renderer logging an instance of `Win32Exception` will result in this output:

`2019-07-19 23:44:53.9813 ERROR  The system cannot find the file specified NativeErrorCode: 2`

**Comment by Giorgi** [comment]
I have a working code implemented in this commit: [Added custom renderer for exceptions.](https://github.com/Giorgi/NLog/commit/e367968f32c5e074060928270e8e4179934b1e21) and I can send a PR.

**Comment by 304NotModified** [comment]
Cool!  PR would be nice!

I think I see a small issue, but it's better to shoe that in a code review

**Comment by Giorgi** [comment]
PR created

**Comment by 304NotModified** [comment]
related, In NLog 4.6.7 `${object-path}` has been added, so you could render properties of an (exception) object.

### Example:
Assume we catch new ExternalException("Exception with errorCode", 5000)
```
${exception:objectpath=HResult} // results in 5000
```

**Comment by Giorgi** [comment]
That's nice but it's weakly typed and easy to make a typo. Strongly typed renderer defined in code would be better and probably faster as it would invoke user provided delegate instead of using reflection

**Comment by snakefoot** [comment]
Work is now being made on opening the access to ObjectReflectionCache. See also #3606

Maybe an idea could be to create an interface that can return the properties for an object-type. Allowing one to override the reflection result (Ex. take the existing result and stripping away certain properties).

Then make ExceptionLayoutRenderer-Formatter that makes use of the ObjectReflectionCache. By default skips the standard properties that can be extracted by the other formatters (Type + Data + Source + StackTrace + etc.)

ExceptionLayoutRenderer will then have a formatter that outputs any additional exception-properties.
ObjectReflectionCache will then have the option to control what properties should be included from an exception-object.

That combination should resolve this issue.

**Comment by snakefoot** [comment]
@Giorgi Created PR #3610 with example of how one can provide own object-reflection-logic.

Then one could combine this with the previous mentioned ExceptionLayoutRenderer-Formatter, that outputs additional/extra exception-properties.

**Comment by snakefoot** [comment]
Resolved with #3610

**Comment by snakefoot** [comment]
@Giorgi NLog 4.7 has been released that includes #3610
